CSS module: generated content (#29316)
mdn/content: files/en-us/web/css/content/index.md

### Values

> The value can be one of two keywords — `none` or `normal` — or, for DOM nodes, the content replacement can be an `<image>`, otherwise a list of of one or more anonymous inline boxes, called a `<content-list>`, that will replace the content of the selected element in the specified order. Each content-list item is either `contents` or of type [`<string>`](#string), [`<image>`](#image), [`<counter>`](#counter), [`<quote>`](#quote), [`<target>`](#target), or [`<leader()>`](#leader), with an optional alternative text value of a `<string>` or `<counter>`:

Author: Changed to:
The value can be one of two keywords — `none` or `normal` — or either `<content-replacement>` when replacing a DOM node or a `<content-list>` for pseudo-elements and margin boxes, with optional alternative text value of a `<string>` or `<counter>`. Content replacement is always an `<image>`. A content-list is a list of of one or more anonymous inline boxes appearing in the order specified. Each content-list item is either `contents` or of type [`<string>`](#string), [`<image>`](#image), [`<counter>`](#counter), [`<quote>`](#quote), [`<target>`](#target), or [`<leader()>`](#leader):

> The value can be one of two keywords — `none` or `normal` — or, for DOM nodes, the content replacement can be an `<image>`, otherwise a list of of one or more anonymous inline boxes, called a `<content-list>`, that will replace the content of the selected element in the specified order. Each content-list item is either `contents` or of type [`<string>`](#string), [`<image>`](#image), [`<counter>`](#counter), [`<quote>`](#quote), [`<target>`](#target), or [`<leader()>`](#leader), with an optional alternative text value of a `<string>` or `<counter>`:

Reviewer: This is an improvement, but it is still hard to parse and work out which of the keywords or data types relates to each of the three options you are talking about.

How about something like this:

```
The value can be:

  - One of two keywords — `none` or `normal`
  - `<content-replacement>` when replacing a DOM node. `<content-replacement>` is always an `<image>`.
  - A `<content-list>` when replacing pseudo-elements and margin boxes, with an optional alternative text value of a `<string>` or `<counter>`.  A content-list is a list of one or more anonymous inline boxes appearing in the order specified. Each `<content-list>` item is either `contents` or of type [`<string>`](#string), [`<image>`](#image), [`<counter>`](#counter), [`<quote>`](#quote), [`<target>`](#target), or [`<leader()>`](#leader).

The keywords and data types mentioned above are described in more detail below:

#### Simple keywords

<!-- List descriptions -->

#### Content replacement data types

<!-- List descriptions -->

####  Content list keywords and data types

<!-- List descriptions -->

```

?

> The value can be one of two keywords — `none` or `normal` — or, for DOM nodes, the content replacement can be an `<image>`, otherwise a list of of one or more anonymous inline boxes, called a `<content-list>`, that will replace the content of the selected element in the specified order. Each content-list item is either `contents` or of type [`<string>`](#string), [`<image>`](#image), [`<counter>`](#counter), [`<quote>`](#quote), [`<target>`](#target), or [`<leader()>`](#leader), with an optional alternative text value of a `<string>` or `<counter>`:

Author: Love it! @chrisdavidmills ready for re-review

    - {{cssxref("counters", "counters()")}}
      - : The {{cssxref("counters", "counters()")}} function also has two forms: 'counters(_name_, _string_)' or 'counters(_name_, _string_, _style_)'. The generated text is the value of all counters with the given name in scope at the given pseudo-element, from outermost to innermost, separated by the specified string. The counters are rendered in the indicated {{cssxref("&lt;list-style-type&gt;")}} (`decimal` by default).

- `<quote>`

  - : The `<quote>` data type includes language- and position-dependent keywords:
    - `open-quote` and `close-quote`
      - : These values are replaced by the appropriate string from the {{cssxref("quotes")}} property.
    - `no-open-quote` | `no-close-quote`
      - : Introduces no content, but increments (decrements) the level of nesting for quotes.

- `<target>` {{Experimental_Inline}}

  - : The `<target>` data type includes three target functions, `<target-counter()>`, `<target-counters()>`, and `<target-text()>` that create cross-reference obtained from the target end of a link. See [Formal syntax](#formal_syntax).

- `<leader()>` {{Experimental_Inline}}

>   - : The `<leader()>` data type inclues a leader function: `leader( <leader-type> )`. This function accepts the keyword values `dotted`, `solid`, or `space` (equal to `leader(".")`, `leader("_")`, and `leader(" ")`, respectively), or a `<string>` as a parameter. When supported and used as a value for `content`, the leader-type provided will be inserted as a repeating pattern visually connecting content across a horizontal line.

Reviewer: ```suggestion
  - : The `<leader()>` data type inclues a leader function: `leader( <leader-type> )`. This function accepts the keyword values `dotted`, `solid`, or `space` (equal to `leader(".")`, `leader("_")`, and `leader(" ")`, respectively), or a `<string>` as a parameter. When supported and used as a value for `content`, the leader-type provided will be inserted as a repeating pattern, visually connecting content across a horizontal line.
```

- `attr(x)`

>   - : The `attr(x)` CSS function retrieves the value of an attribute of the selected element or pseudo-element's originating element. The value of the element's attribute `x` as an unparsed string. If there is no attribute `x`, an empty string is returned. The case-sensitivity of attribute name parameter depends on the document language.

Reviewer: ```suggestion
  - : The `attr(x)` CSS function retrieves the value of an attribute of the selected element, or the pseudo-element's originating element. The value of the element's attribute `x` is an unparsed string representing the attribute name. If there is no attribute `x`, an empty string is returned. The case sensitivity of the attribute name parameter depends on the document language.
```

> - alternative text `/ <string> | <counter>`

Reviewer: ```suggestion
- alternative text: `/ <string> | <counter>`
```

Not 100% sure if this is the best way, but when I originally read this it took me a while to get my head around the fact that the slash is part of the syntax. I thought some kind of separator might help.

>   - : Alternative text may be specified for an image or any `<content-list>` items, by appending a forward slash and then a string of text or counter. The alternative text is intended for speech output by screen-readers, but may also be displayed in some browsers. Note that if the browser does not support alternative text, the `content` declaration will be considered invalid and will be ignored. The {{cssxref("string", "/ &lt;string>")}} or {{cssxref("counter", "/ &lt;counter>")}} specifies the "alt text" for the element.

Reviewer: ```suggestion
  - : Alternative text may be specified for an image or any `<content-list>` items, by appending a forward slash and then a string of text or a counter. The alternative text is intended for speech output by screen-readers, but may also be displayed in some browsers. Note that if the browser does not support alternative text, the `content` declaration will be considered invalid and will be ignored. The {{cssxref("string", "/ &lt;string>")}} or {{cssxref("counter", "/ &lt;counter>")}} data types specify the "alt text" for the element.
```

## Examples

> The first examples created generated content on pseudo-elements. The last three are [examples of element replacement](#element-replacement-with-url).

Reviewer: ```suggestion
The first five examples create generated content on pseudo-elements. The last three are [examples of element replacement](#element-replacement-with-url).
```

### Appending strings based on an element's class

> This example inserts generated text, coloring it red, after the text based on an element's class name.

Reviewer: ```suggestion
This example inserts generated text after the text of elements that have a particular class name. The text is colored red.
```

q::after {
  content: close-quote;
}
```

#### Result

{{EmbedLiveSample('Quotes', '100%', 200)}}

> Note the [type of quotes generated](/en-US/docs/Web/CSS/quotes#auto_quotes) is based on the language. Browsers add open- and close-quotes before and after {{HTMLElement("q")}} elements by default, so the quotes in this example would appear without them being explicitly set. They could have been turned off by setting `no-open-quote` and `no-close-quote`, or `none` value. They can also be turned off by setting the {{cssxref("quotes")}} property to `none` instead.

Reviewer: ```suggestion
Note the [type of quotes generated](/en-US/docs/Web/CSS/quotes#auto_quotes) is based on the language. Browsers add open- and close-quotes before and after {{HTMLElement("q")}} elements by default, so the quotes in this example would appear without them being explicitly set. They could have been turned off by setting the respective `content` property values to `no-open-quote` and `no-close-quote`, or by setting them both to `none`. They can also be turned off by setting the {{cssxref("quotes")}} property to `none` instead.
```

### Strings with attribute values

This example is useful for print stylesheets. It uses an [attribute selector](/en-US/docs/Web/CSS/Attribute_selectors) to select every fully qualified secure link, adding the value of the `href` attribute after the link text as the content of the {{cssxref("::after")}} pseudo-element.

#### HTML

```html
<ul>
  <li><a href="https://mozilla.com">Mozilla</a></li>
  <li><a href="/">MDN</a></li>
  <li><a href="https://openwebdocs.org">OpenWebDocs</a></li>
</ul>
```

#### CSS

```css
a[href^="https://"]::after
{
  content: " (URL: " attr(href) ")";
  color: darkgreen;
}
```

#### Result

{{EmbedLiveSample('Strings_with_attribute_values', '100%', 200)}}

The generated content is the value of the `href` attribute, prepended by "URL: ", with a space, all in parentheses.

### Adding an image with alternative text

> This example inserts an image before the link. Alternative text that a screen reader can output as speech in included,

Reviewer: ```suggestion
This example inserts an image before the link. Alternative text that a screen reader can output as speech is included,
```

> which some browsers may display. To indicate if alternative text after a content list is supported or not, a fallback `content`, the CSS includes the message " - no alt text - ".

Reviewer: This sentence is difficult to parse, and I'm not 100% sure what you are trying to say. Can you please clarify it?

> This example inserts an image before all links. Two `content` values are provided. The later `content` value includes an image with alternative text that a screen reader can output as speech. If a browser does not support alternative text, this declaration will be considered invalid, with the previous `content` value displaying. This fallback content list include an image and the message " - alt text is not supported - ".

Reviewer: ```suggestion
This example inserts an image before all links. Two `content` values are provided. The later `content` value includes an image with alternative text that a screen reader can output as speech. If a browser does not support alternative text, this declaration will be considered invalid, with the previous `content` value displaying. This fallback content list includes an image and the message " - alt text is not supported - ".
```

#### HTML

```html
<a href="https://www.mozilla.org/en-US/">Mozilla Home Page</a>
```

#### CSS

The CSS to show the image and set the alternative text is shown below.
This also sets the font and color for the content.
This will be used on browsers that _display_ the alternative text and in browsers that don't support alternative text and show the the fallback `content` value.

```css
a::before {
  /* fallback content */
  content: url("https://mozorg.cdn.mozilla.net/media/img/favicon.ico")
    " - no alt text - ";
  /* content with alternative text */
  content: url("https://mozorg.cdn.mozilla.net/media/img/favicon.ico") /
    " MOZILLA: ";
  font:
    x-small Arial,
    sans-serif;
  color: gray;
}
```

#### Result

{{EmbedLiveSample('Adding_an_image_with_alternative_text', '100%', 60)}}

If using a screen reader, it should speak the word "MOZILLA" when it reaches the image. If supported (if the "no alt text" is not showing), you can select the `::before` pseudo-element with your developer tools selection tool, and view the {{glossary("accessible name")}} in the accessibility panel.

> Note that on a browser that does not support the alternative text syntax, the whole line is invalid.

Reviewer: ```suggestion
Note that in browsers that don't support the alternative text syntax, the whole line is invalid.
```

Also, maybe put it in a note box if it is a note?

Author: didn't put a note around, but did add an explanatory sentence.
> In browsers that don't support the alternative text syntax the whole declaration containing the alt text is invalid. In this case, the previous `content` value will be used, showing the image and "alt text is not supported" text.

In this case, the previous `content` value, if supported, will be used.

### Element replacement with `url()`

> This example replaces a regular element! The element's contents are replaced with an SVG using the {{cssxref("url", "url()")}} image function. We include generated content, attempting to add the `id` of the element generated on `::after`. This pseudo-element is not generated in browsers that support content replacement as the element is replaced.

Reviewer: Should "that support content replacement" be "that don't support content replacement"? The sentence sounds a bit contradictory as-is.

Author: It's correct. Pseudo-elements aren't rendered on replaced elements. If the element is replaced, the ::after and ::before will not be generated. Tried to explain it better.

Reviewer: Yup, makes sense now
